Allow specifying if a metric is ascending or descending

punchagan opened the pull request:
The ascending field in the metric metadata is used to indicate whether an
increase in the value is good/bad. When the metric is ascending and the value
increases, or the metric is descending and the value decreases, the delta value
is displayed in green. If not, the delta is displayed in red.

![image](https://user-images.githubusercontent.com/315678/147069274-8f8e20ba-fd12-4def-b1ab-ee93d7d85e49.png)

art-w commented on frontend/src/BenchmarkTest.res:
```diff
@@ -68,6 +80,11 @@ let renderMetricOverviewRow = (
       }
+      let color = switch row.delta {
+      | Some(delta) =>
+        (delta >= 0. && row.ascending) || (delta <= 0. && !row.ascending) ? Sx.green300 : Sx.red300
```
The equality `( = )` operator can be used on boolean to mean "both or neither" : `(delta >= 0.) = row.ascending` (... it's a fun trick, but it probably makes the code harder to read if you don't know it, so feel free to leave the explicit condition!)

punchagan replied:
This is neat trick, that's only slightly harder to read. I'll switch over.

art-w commented on pipeline/lib/current_bench_json.ml:
```diff
@@ -131,11 +134,20 @@ module V2 = struct
   let metric_of_json t =
     let units = Json.get "units" t |> Json.to_string_option |> default "" in
+    let ascending =
+      Json.get "ascending" t |> Json.to_bool_option |> default true
```
I'm not sure that `ascending` is the most clear term, does `true` indicates that higher-is-better? (by default I think that lower-is-better is the most common type of benchmarks)

punchagan replied:
Yes, I wasn't sure that `ascending` is the best term to use either. I took the term from a [measurement and evaluation framework](https://iatistandard.org/en/iati-standard/203/activity-standard/iati-activities/iati-activity/result/indicator/) I worked with in the past.  I'm open to other suggestions for the name. 

I haven't looked at all the benchmarks we currently have, but for some thing like the compiler, I think `lower-is-better` would be more common. 

Also, I'm not sure if we should have a default value/meaning for the existing benchmarks that have already been written. It could be confusing for users who aren't aware of the new `flag` that we are adding here?

art-w replied:
Yes it could be a `bool option` for when the flag is not set (or an explicit sum type to improve readability: `Unknown | Lower_is_better | Higher_is_better`... sometimes it's not clear what a bool means!)
But it seems easy enough to fix by the user if the default is not right... And we could help by using the metric to guess a better default? (a simple heuristic, as ".../s" is more probably a HiB, but otherwise benchmarks are generally LiB ?)

The word `ascending` makes sense with the documentation! I'll try to find a more explicit word just in case :)

art-w commented:
As usual this is excellent! I'm just not sure about the word `ascending` in the json (but I don't have a great suggestion either to evoke higher-is-better/lower-is-better)

The feature could be displayed even more prominently in the frontend! (but this can be done in follow up PRs) For example, it would be nice to have the info near the graphs of the "good" direction to facilitate reading? (as an arrow or icon if space is an issue?)

Another random idea: On mixed metrics (many lower and higher is better graphs), perhaps it is hard to switch between the modes of reading (a bit like a survey where some questions are randomly negated and you can't just assume that "yes" is the good answer). We could choose that "lower-is-always-better" and reverse the sign of the higher-is-better graphs so that reading all graphs becomes "are all the metrics visually decreasing?" (... this is probably a bit surprising and would require some works on the graphs axis to explain!)

punchagan commented:
> As usual this is excellent! I'm just not sure about the word `ascending` in the json (but I don't have a great suggestion either to evoke higher-is-better/lower-is-better)

Thanks for the review, @art-w. :)  @gs0510 would you have any thoughts on using `ascending` to indicate `higher-is-better` metrics? I stole it from this [M&E framework](https://iatistandard.org/en/iati-standard/203/activity-standard/iati-activities/iati-activity/result/indicator/#attributes) I used in the past. 
 
> The feature could be displayed even more prominently in the frontend! (but this can be done in follow up PRs) For example, it would be nice to have the info near the graphs of the "good" direction to facilitate reading? (as an arrow or icon if space is an issue?)

I agree on making the feature more prominent in the UI. And I punted on it, for further discussion or ideas on this PR, or for future PRs. I'll wait to see if anyone else has any ideas for this too. /cc @gs0510

gs0510 commented:
I agree that ascending (=true) is a bit confusing for `higher-is-better`. Can we just call it `higher-is-better` (it doesn't sound as standard as `ascending` though).

Making it more prominent on the UI sounds good! :)) (I don't know if you want specific ideas, but starting with an arrow next to the graph sounds good :) )

art-w commented:
Perhaps the bool is causing most of the confusion? Would a string be better?

`"ascending": true,`
vs
`"direction": "ascending",` (or `"trend"` for the key?... or `"higher"` and `"lower"` for the value)

gs0510 commented:
Ah yes, the bool is the cause of the confusion (always 😁), a string type would definitely be better :))

punchagan commented:
![image](https://user-images.githubusercontent.com/315678/147333996-906178e5-c982-4cfc-b1f3-1a6b258d6f7f.png)

I've added the color indicator to the subtitle on the graphs, and also added some up/down arrows.  The up and down arrows are tracking the actual deltas, while the color is based on the delta and whether the metric is higher/lower is better. I don't know if the arrows are confusing. Happy to take feedback on this. 

Also, switched to a string parameter --`"trend": "higher-is-better"`

gs0510 commented:
Ah nice work! The arrows are indeed confusing 😿 I think it's the problem of double negatives, because there's color + arrow direction + (+/-) so it's hard to figure out what it actually means. 😬

gs0510 reviewed (approved):
LGTM! Thanks @punchagan, I think the idea of red/green when metrics are better/worse depending on higher-is-better/lower-is-better is nice :))

gs0510 commented:
Feel free to merge this after rebase @punchagan  :)

punchagan commented:
This PR (and probably `main` before I rebased this PR against `main`?) had a problem with running `make bench` because of using spaces instead of tabs in the `Makefile`.  I pushed a commit to `main` to fix this. It would be useful to continue to run `make bench` as part of the CI which we used to previously. 
@gs0510 did we remove this intentionally? Or is something wrong somewhere?
